### ISSUE TITLE
Feature: directives capturing TemplateRef

### ISSUE THREAD

**Issue by pkozlowski-opensource** [angular/angular#7624]
This is part of the discussion happening in #7546, but this issue only captures one aspect of it suggested by @tbosch : https://github.com/angular/angular/issues/7546#issuecomment-195498979

## Problem

Let's consider a directive that captures a reference to a `TemplateRef`. To make the discussion more concrete we can discuss `menu-item` from this example (complete one in plunker: http://plnkr.co/edit/sleVVrgqo66JN9G48yyj?p=preview): 

```HTML
<menu>
    <template menu-item>A<template>
    <template menu-item>B<template>
</menu>
``` 

Here `menu-item` needs to capture `TemplateRef` and this is why it needs to be placed on the `<template>` element. But this is not the API we would like to offer to end-users! What we would like to have for directive users is:

```HTML
<menu>
  <menu-item>A</menu-item>
  <menu-item>B</menu-item>
</menu>
```

## Feature request

Make it possible to mark a directive as "template capturing" so its content (here: `A` or `B`) would be accessible, from JS side, as `TemplateRef`. This would make directive's content available (as `TemplateRef`) without the need to use `<template>` elements explicitly.

At the end of the day what is needed what is a mechanism to capture and access, from JS, content of directive's childrens.

**Comment by pkozlowski-opensource** [comment]
Additional background / motivation / philosophy: when discussing the topic with Tobias we've been trying to categorize how children's content is accessed / projected. IMO we've got 3 slightly different use-cases:

1. Some (or all) children are nodes that we want to statically project. As of today this use-case  is covered by `<ng-content>`. Please note that projected content doesn't be on / wrapped by a `<template>` tag. It also doesn't get local variables from a directive that is projecting children. A given chunk can be projected only once.

2. Some children are used as a **template** for a directive that does "manual" projection. NgFor is a typical case but there are plenty of others (ex.: supply a template to an auto-complete widget - such template can be used to display each item in the result list). As of today pieces of HTML that acts as a "template" needs to be wrapped in a `<template>` tag and quite often need local variables from a projecting directive. A given template is projected 1 or multiple times.

3. Similar to (1) but where we would like to do "manual" projection (`menu` and `menu-item` from our example). This case is closer to (1) as given child is projected at most one time and doesn't need local variables, but needs power of (2).

The case (3) is the one I'm after here. It requires more power than currently offered by static `<ng-content>` (ex.: ability to filter out children imperatively, in JS). Technically speaking we could use (2) but from the user's perspective it doesn't make sense to use `<template>` tags - user's don't supply templates that are going to be used to stamp out things. Rather they supply content to be projected. It just happens that a directive needs more powerful projection mechanism.

**Comment by pkozlowski-opensource** [comment]
This will be better fixed by #8563
